src/neuron/cache/mechanism_range.hpp

```diff
@@ -0,0 +1,256 @@
+     * @f$\texttt{i}^{th}@f$ RANGE variable.
+     * @see container::detail::field_data<Tag, true>::data_ptrs()
+     */
+    double* const* m_data_ptrs{};
+
+    /**
+     * @brief Pointer to a range of array dimensions for the RANGE variables in this mechanism.
+     *
+     * @c m_data_array_dims[i] is the array dimension of the @f$\texttt{i}^{th}@f$ RANGE variable.
+     */
+    int const* m_data_array_dims{};
+
+    /**
+     * @brief Pointer to a range of pointers to the start of POINTER variable caches.
+     *
+     * @c m_pdata_ptrs[i][j] is the @c double* corresponding to the @f$\texttt{i}^{th}@f$ @c pdata /
+     * @c dparam field and the @f$\texttt{j}^{th}@f$ instance of the mechanism in the program.
+     * @see MechanismInstance::MechanismInstance(Prop*) and @ref nrn_sort_mech_data.
+     */
+    double* const* const* m_pdata_ptrs{};
+
+    /**
+     * @brief Offset of this contiguous range of mechanism instances into the global range.
+     *
+     * Typically if there is more than one thread in the process then the instances of a particular
+     * mechanism type will be distributed across multiple NrnThread objects and processed by
+     * different threads, and the mechanism data will be permuted so that the instances owned by a
+     * given thread are contiguous. In that case the MechanismRange for the 0th thread would have an
+     * @c m_offset of zero, and the MechanismRange for the next thread would have an @c m_offset of
+     * the number of instances in the 0th thread.
+     *
+     * @see @ref nrn_sort_mech_data.
+     */
+    std::size_t m_offset{};
+};
+/**
+ * @brief Specialised version of MechanismRange for a single instance.
+ *
+ * This is used inside generated code that takes a single mechanism instance (Prop) instead of a
+ * range of instances (Memb_list). A key feature of methods that take Prop is that they should
+ * *not* require a call to nrn_ensure_model_data_are_sorted(). This is conceptually fine, as if
+ * we are only concerned with a single mechanism instance then it doesn't matter where it lives
+ * in the global storage vectors. In this case, @ref m_dptr_cache contains an array of pointers
+ * that @ref m_dptr_datums can refer to.
+ */
+template <std::size_t NumFloatingPointFields, std::size_t NumDatumFields>
+struct MechanismInstance: MechanismRange<NumFloatingPointFields, NumDatumFields> {
+    /**
+     * @brief Shorthand for the MechanismRange base class.
+     */
+    using base_type = MechanismRange<NumFloatingPointFields, NumDatumFields>;
+
+    /**
+     * @brief Construct from a single mechanism instance.
+     * @param prop Handle to a single mechanism instance.
+     */
+    MechanismInstance(Prop* prop)
+        : base_type{_nrn_mechanism_get_type(prop), mechanism::_get::_current_row(prop)} {
+        if (!prop) {
+            // grrr...see cagkftab test where setdata is not called(?) and extcall_prop is null(?)
```

Comment:
Maybe @nrnhines could check/comment. The stuff with setdata and extcall_prop seems very niche and not widely used/tested.

Reply:
I'm not familiar with this context. Wrote the following but decided it really was beside the point. Will have to tutored on what this is about.
setdata was needed to call a density mechanism function from hoc if the mechanism function referred to range variables. (there was no corresponding problem with POINT_PROCESS since a function call was prefixed by the specific object instance involved).

This might be a non sequitor but I believe that in Old NEURON, lines involving prop often referred to HOC defined mechanisms (rarely used, and if so, probably only by me). There are two functions in HOC called make_mechanism() and make_pointprocess().
The only example of use I can find is nrntest/nrniv/makemech.hoc
which creates a mechanism called Max that records the maximum value of voltage during a run. It certainly can have multiple instances.
The implementation of all that is in nrn/src/nrniv/hocmech.cpp